docs/nima/testing.adoc

```diff
@@ -0,0 +1,393 @@
+
+include::{rootdir}/includes/dependencies.adoc[]
+[source,xml]
+----
+<dependency>
+    <groupId>io.helidon.nima.testing.junit5</groupId>
+    <artifactId>helidon-nima-testing-junit5-webserver</artifactId>
+    <scope>test</scope>
+</dependency>
+----
+
+== Usage
+
+Helidon provides a rich set of extensions based on JUnit 5 for Helidon Níma testing. Testing can be done with automatic server start-up, configuration, and shutdown. Testing can also be done without full server start-up with `DirectClient` when no real sockets are created.
+
+== API
+
+There are two main annotations that you can use to test the Níma server.
+
+* `@ServerTest` is an integration test annotation that starts the server (opens ports) and provides client injection pre-configured for the server port(s).
+* `@RoutingTest` is a unit test annotation that does not start the server and does not open ports but provides a direct client (with the same API as the usual network client) to test routing.
+
+The additional annotation `@Socket` can be used to qualify the injection of parameters into test constructors or methods, such as to obtain a client configured for the named socket.
+
+The following table lists the supported types of parameters for the `@SetUpRoute` annotated methods. Such methods MUST be static
+and may have any name. The `@SetUpRoute` annotation has `value` with socket name (to customize the setup for a different socket).
+
+- Parameter type - supported class of a parameter
+- Annotation - which annotations support this parameter
+- Modules - which Níma extension modules support this signature
+
+.Parameters for the `@SetUpRoute` annotated methods.
+|===
+|Parameter Type |Annotation |Modules |Notes
+
+|`HttpRouting.Builder`
+|`@ServerTest`, `@RoutingTest`
+|
+|
+
+|`HttpRules`
+|`@ServerTest`, `@RoutingTest`
+|
+|Same as `HttpRouting.Builder`, only routing setup
+
+|`Router.RouterBuilder<?>`
+|`@ServerTest`, `@RoutingTest`
+|
+|
+
+|`SocketListener.Builder`
+|`@ServerTest`
+|
+|
+
+|`WebSocketRouting.Builder`
+|`@ServerTest`, `@RoutingTest`
+|websocket
+|
+
+|===
+
+In addition, a static method annotated with `@SetUpServer` can be defined for `@ServerTest`, which has a single parameter of `WebServer.Builder`.
+
+The following table lists the injectable types (through constructor or method injection).
+
+- Type - type that can be injected
+- Socket? - if checked, you can use the `@Socket` annotation to obtain a value specific to that named socket
+- Annotation - which annotations support this injection
+- Modules - which Níma extension modules support this injection
+- Notes - additional details
+
+.Injectable types.
+|===
+|Type |Socket? |Annotation |Modules |Notes
+|`WebServer`
+|
+|`@ServerTest`
+|
+|Server instance (already started)
+
+|`URI`
+|x
+|`@ServerTest`
+|
+|URI pointing to a port of the webserver
+
+|`SocketHttpClient`
+|x
+|`@ServerTest`
+|
+|The client that allows sending of anything, to test bad requests and such.
```

Comment:
This client allows you to send anything in order to test for bad requests or other issues.

Reply:
Fixed